tests/test_concurrent_map.cpp

```diff
@@ -0,0 +1,333 @@
+  });
+
+  auto second = std::thread([&]() {
+    for (auto n : second_numbers) {
+      skip_list.Erase(n);
+    }
+  });
+
+  first.join();
+  second.join();
+
+  if (!skip_list.Contains(num)) {
+    REQUIRE(skip_list.Insert(num, num));
+  } else {
+    REQUIRE(skip_list.Erase(num));
+  }
+}
+
+TEST_CASE("One thread inserts and second check for contain the same element",
+          "[Concurrency]") {
+  auto skip_list = SL<int, int>{};
+
+  int num = 10;
+
+  auto first_numbers = chunk(kThousand, random(num, num)).get();
+  auto second_numbers = chunk(kThousand, random(num, num)).get();
+
+  auto first = std::thread([&]() {
+    for (auto n : first_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  auto second = std::thread([&]() {
+    for (auto n : second_numbers) {
+      skip_list.Contains(n);
+    }
+  });
+
+  first.join();
+  second.join();
+
+  REQUIRE(skip_list.Contains(num));
+}
+
+TEST_CASE(
+    "Tho threads insert different numbers, then four threads check for contain",
+    "[Concurrency]") {
+  auto skip_list = SL<int, int>{};
+
+  auto first_write_numbers =
+      chunk(10 * kThousand, random(0, kThousand - 1)).get();
+  auto second_write_numbers =
+      chunk(10 * kThousand, random(kThousand, 2 * kThousand)).get();
+
+  auto first_write_size = static_cast<int>(first_write_numbers.size());
+  auto second_write_size = static_cast<int>(second_write_numbers.size());
+
+  auto first_read_numbers =
+      std::vector<int>(std::begin(first_write_numbers),
+                       std::begin(first_write_numbers) + first_write_size / 2);
+  auto second_read_numbers = std::vector<int>(
+      std::begin(first_write_numbers) + first_write_size / 2 + 1,
+      std::end(first_write_numbers));
+
+  auto third_read_numbers = std::vector<int>(
+      std::begin(second_write_numbers),
+      std::begin(second_write_numbers) + second_write_size / 2);
+  auto fourth_read_numbers = std::vector<int>(
+      std::begin(second_write_numbers) + second_write_size / 2 + 1,
+      std::end(second_write_numbers));
+
+  auto first_write = std::thread([&]() {
+    for (auto n : first_write_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+  auto second_write = std::thread([&]() {
+    for (auto n : second_write_numbers) {
+      skip_list.Insert(n, n);
+    }
+  });
+
+  first_write.join();
+  second_write.join();
+
+  auto first_read = std::thread([&]() {
+    for (auto n : first_read_numbers) {
+      REQUIRE(skip_list.Contains(n));
+    }
+  });
+  auto second_read = std::thread([&]() {
+    for (auto n : second_read_numbers) {
+      REQUIRE(skip_list.Contains(n));
+    }
+  });
+  auto third_read = std::thread([&]() {
+    for (auto n : third_read_numbers) {
+      REQUIRE(skip_list.Contains(n));
+    }
+  });
+  auto fourth_read = std::thread([&]() {
+    for (auto n : fourth_read_numbers) {
+      REQUIRE(skip_list.Contains(n));
+    }
+  });
+
+  first_read.join();
+  second_read.join();
+  third_read.join();
+  fourth_read.join();
```

Comment:
In case of more than two threads, it's reasonable to have a vector of threads, so you don't have to repeat yourself.